# 1. Algorithm
[1028. Convert to Base -2](https://leetcode.com/contest/weekly-contest-130/problems/convert-to-base-2/)
```Java        
class Solution {
    public String baseNeg2(int N) {
        if(N == 0) {
            return "0";
        }
        StringBuilder sb = new StringBuilder();
        int base = -2;
        while(N != 0) {
            int remainder = N % base;
            N /= base;
            if(remainder < 0) {
                remainder += (-base);
                N += 1;
            }
            sb.append(remainder);
        }
        return sb.reverse().toString();
    }
}

```
[1030. Next Greater Node In Linked List](https://leetcode.com/contest/weekly-contest-130/problems/next-greater-node-in-linked-list/)
```
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        Stack<Integer> stk = new Stack<>();
        List<Integer> list = new ArrayList<>();
        while(head != null) {
            list.add(head.val);
            head = head.next;
        }
        int[] res = new int[list.size()];
        
        for(int i = list.size() - 1; i >= 0; i--) {
            while(!stk.isEmpty() && stk.peek() <= list.get(i)) {
                stk.pop();
            }
            res[i] = stk.isEmpty() ? 0 : stk.peek();
            stk.push(list.get(i));
        }
        return res;
    }
}
```
# 2. Review
## Java Exception
This week, I read some articles about Java exception, which I have been using for a long time but am actually not very familiar with.

### Exception Types
All exception and errors extend from a common `java.lang.Throwable` parent class.
There are some classes that have special meaning:
  1. java.lang.Error - JVM Error
  2. java.lang.RuntimeException - Programming Error
  3. java.lang.Exception - System Error

### java.lang.RuntimeException vs java.lang.Exception
1. java.lang.RuntimeException </br>
   `RuntimeException`s are exceptions that can be prevented programmatically. `NullPointerException` and `ArrayIndexOutOfBoundException` are commonly seen RuntimeExceptions. Compilers DO NOT check RuntimeException.

2. java.lang.Exception </br>
   For any exception that derives from `Exception`, we should put a try/catch block around or add a `throws` clause to the method.

### Checked vs Unchecked Exceptions
1. Unchecked Exception </br>
   `Errors` and `RuntimeExceptions` are `unchecked exception` - the compiler does not enforce us to handle them explicitly. </br>
   It is assumed that the application cannot do anything to recover from these exceptions (at runtime). </br>
   All other Exceptions are checked  —  that is, the compiler enforces that you handle them explicitly. </br>

2. Checked Exception </br>
   Methods that generate checked exceptions must declare that they throw them. </br>
   Methods that invoke other methods that throw checked exceptions must either handle them (they can be reasonably expected to recover) or let them propagate by declaring that they throw them. </br>

### Exception Lifecycle
  1. After an exception is created, it is handed off to the runtime system.
  2. If the runtime can find a handler for the exception by backtracking the call stack, then the exception will be handled or re-thrown.
  3. If the runtime can not find a handler, the exception stack trace will be printed to the standard error channel(stderr) and the application will abort execution.

### References
  1. [Java Fundamentals Tutorial: Exceptions](https://www.protechtraining.com/bookshelf/java_fundamentals_tutorial/exceptions)
  2. [Java Catch Multiple Exceptions, Rethrow Exception](https://www.journaldev.com/629/java-catch-multiple-exceptions-rethrow-exception)
  
  
# 3. Tip
1. [Convert JSON to Map<String, Object>](https://stackoverflow.com/questions/443499/convert-json-to-map)
2. An easy way to convert List<Integer> to int[]: </br>
   int[] arr = list.stream().mapToInt(i -> i).toArray();
3. Learned some basic usage of Optional in Java 8.
```
package com.Java8;

import java.util.List;
import java.util.Optional;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

public class OptionalStudy {
    private Optional<String> getEmpty() {
        return Optional.empty();
    }

    private Optional<String>  getHello() {
        return Optional.of("hello");
    }

    private Optional<String> getBye() {
        return Optional.of("bye");
    }

    public void testPresent() {
        Optional<String> empty = Optional.empty();
        assertFalse(empty.isPresent());

        // orElse orElseGet orElseThrow
        String nullRef = null;
        // Optional<String> optStr = Optional.of(nullRef); // this will throw NullPointerException
        Optional<String> optStr1 = Optional.ofNullable(nullRef);
        String str1 = Optional.ofNullable(nullRef).orElse("Another Value");
        String str2 = Optional.ofNullable(nullRef).orElseGet(() -> "Another Value");
        // String str3 = Optional.ofNullable(nullRef).orElseThrow(
        //       IllegalArgumentException::new);
    }

    public void testMap() {
        List<String> list = Stream.of(getEmpty(), getHello(), getBye())
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
        // class java.util.ArrayList
        System.out.println(list.getClass());


        Optional<Person> person = Optional.of(new Person());
        Optional<String> optName = person.flatMap(Person::getName);
        System.out.println(optName);

    }

    public void testChainingOptionals() {
        // chaining Optionals
        String defaultStr = Stream.<Supplier<Optional<String>>>of(
                () -> Optional.empty(),
                () -> Optional.empty())
                .map(Supplier::get)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst()
                .orElseGet(() ->  "default");

        System.out.println(defaultStr);
    }

    public static void main(String[] args) {
        OptionalStudy os = new OptionalStudy();

        os.testPresent();
        os.testMap();
        os.testChainingOptionals();
    }

    private static class Person {
        String name;

        public Person() {
            name = "phil";
        }
        public Optional<String> getName() {
            return Optional.of(name);
        }
    }
}
```

# 4. Share
### [Why you Procrastinate(It Has Nothing to Do With Self-Control)](https://www.nytimes.com/2019/03/25/smarter-living/why-you-procrastinate-it-has-nothing-to-do-with-self-control.html)
In this article, the author says that procrastination is not laziness, it's more about emotion. Procrastination is an emotion regulation problem, not a time management problem. Procrastination can be understood as "the primacy of short-term mood repair over the longer-term pursuit of intended actions".
We all know that procrastination has bad effect. But we engage in the irrational cycle of chronic procrastination because of an inability to manage negative moods around a task. That we get momentary relief from procrastination is what makes the cycle especially vicious. 

We are designed to focus on the present rather than the future. On a neural level, we perceive our "future selves" more like strangers than as parts of ourselves. So when we procrastinate, our brains actually think that the tasks are somebody else's problem.

To deal with procrastination, we should find a better reward than avoidance: </br>
  1. Forgive yourself in the moments you procrastinate
  2. Another tactic is the related practice of Self-Compassion, which means treating ourselves with kindness and understanding in the face of our mistakes and failures.
    
Some other healthier ways to manage the feeling that typically trigger procrastination:
  1. Cultivate curiosity. 
  2. Consider the next action, don't wait to be in the mood to do a certain task.
  3. Make your temptations more inconvenient like delete all attractive apps on your phone if you compulsively check social media.
